msi-ec.c

```diff
@@ -425,38 +701,35 @@ static ssize_t fan_mode_show(struct device *device,
 	u8 rdata;
 	int result;
 
-	result = ec_read(MSI_EC_FAN_MODE_ADDRESS, &rdata);
+	result = ec_read(conf->fan_mode.address, &rdata);
 	if (result < 0)
 		return result;
 
-	switch (rdata) {
-	case MSI_EC_FAN_MODE_AUTO:
-		return sprintf(buf, "%s\n", "auto");
-	case MSI_EC_FAN_MODE_BASIC:
-		return sprintf(buf, "%s\n", "basic");
-	case MSI_EC_FAN_MODE_ADVANCED:
-		return sprintf(buf, "%s\n", "advanced");
-	default:
-		return sprintf(buf, "%s (%i)\n", "unknown", rdata);
+	for (int i = 0; i <= conf->fan_mode.max_mode; i++) {
+		if (rdata == conf->fan_mode.mode_values[i]) {
+			return sprintf(buf, "%i\n", i);
+		}
 	}
+
+	return sprintf(buf, "%s (%i)\n", "unknown", rdata);
 }
 
 static ssize_t fan_mode_store(struct device *dev, struct device_attribute *attr,
 			      const char *buf, size_t count)
 {
-	int result = -EINVAL;
-
-	if (streq(buf, "auto"))
-		result =
-			ec_write(MSI_EC_FAN_MODE_ADDRESS, MSI_EC_FAN_MODE_AUTO);
+	u8 wdata;
+	int result;
 
-	if (streq(buf, "basic"))
-		result = ec_write(MSI_EC_FAN_MODE_ADDRESS,
-				  MSI_EC_FAN_MODE_BASIC);
+	result = kstrtou8(buf, 10, &wdata);
+	if (result < 0)
+		return result;
 
-	if (streq(buf, "advanced"))
-		result = ec_write(MSI_EC_FAN_MODE_ADDRESS,
-				  MSI_EC_FAN_MODE_ADVANCED);
+	if (wdata > conf->fan_mode.max_mode) {
+		result = -EINVAL;
+	} else {
+		result = ec_write(conf->fan_mode.address,
+				  conf->fan_mode.mode_values[wdata]);
+	}
```

Comment:
4. See my earlier comment for the shift mode. As user, I need know if my laptop is using basic fan mode (1 threshold level) or advanced fan mode (6 threshold levels). Actually, no advanced levels were implemented in the codebase, just the basic level (`/sys/devices/platform/msi-ec/cpu/basic_fan_speed`) was. Other community projects do implement both basic/advanced levels. However, don't think about it now because exceeds the scope of this pull request.

Reply:
This is similar to shift modes. This repo doesn't have a silent mode but msi-ec-modern does. So either you just didn't implement it or your laptop doesn't have it at all and supported fan modes should be provided by configurations

Reply:
Made fan modes human readable. I'm assuming all laptops have the same fan modes for now

Reply:
My laptop don't have silent mode, just auto/basic/advanced modes. **Maybe we need an `available_fan_modes` here**. For the rest, I read your code and looks good for me. I won't close this comment because I think that available_fan_modes is something important before finishing.

Reply:
Done